leptonai/util/util.py

```diff
@@ -59,19 +61,26 @@ def patch(obj, attr, val):
         setattr(obj, attr, old_val)
 
 
-def asyncfy(func):
+def asyncfy(func, semaphore: Optional[anyio.Semaphore]):
     """Decorator that makes a function async
 
     Args:
         func (function): Function to make async
+        semaphore (anyio.Semaphore): Semaphore to use for concurrency control. If func
+            is already async, this argument is ignored. If func is sync, it will be
+            guarded by the semaphore.
     """
 
     if inspect.iscoroutinefunction(func):
         return func
 
     @wraps(func)
     async def async_func(*args, **kwargs):
-        return func(*args, **kwargs)
+        if semaphore is None:
```

Comment:
nit: there is `contextlib.nullcontext()` we can use as the default value of `semphore` arg of `asyncfy`, and then we can unconditionally do `with semphore:`

Reply:
Thanks! Probably keeping it this way is a bit easier, because type hint `Optional[Semaphore]` would be a bit more clear than `Union[Semaphore, contextlib.nullcontext]`...